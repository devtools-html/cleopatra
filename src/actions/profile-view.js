/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import type {
  ProfileSelection, CallTreeFilter, ImplementationFilter,
} from '../types/actions';
import type { Action, ThunkAction } from '../types/store';
import type { ThreadIndex, IndexIntoFuncTable, IndexIntoMarkersTable } from '../types/profile';

/**
 * The actions that pertain to changing the view on the profile, including searching
 * and filtering. Currently the call tree's actions are in this file, but should be
 * split apart. These actions should most likely affect every panel.
 */
export function changeSelectedFuncStack(
  threadIndex: ThreadIndex,
  selectedFuncStack: IndexIntoFuncTable[]
): Action {
  return {
    type: 'CHANGE_SELECTED_FUNC_STACK',
    selectedFuncStack, threadIndex,
  };
}

export function changeSelectedThread(selectedThread: ThreadIndex): Action {
  return {
    type: 'CHANGE_SELECTED_THREAD',
    selectedThread,
  };
}

export function changeThreadOrder(threadOrder: ThreadIndex[]): Action {
  return {
    type: 'CHANGE_THREAD_ORDER',
    threadOrder,
  };
}

export function hideThread(
  threadIndex: ThreadIndex,
  threadOrder: ThreadIndex[],
  hiddenThreads: ThreadIndex[]
): ThunkAction<void> {
  return dispatch => {
    // Do not allow hiding the last thread.
    if (hiddenThreads.length + 1 === threadOrder.length) {
      return;
    }

    dispatch(({
      type: 'HIDE_THREAD',
      threadIndex,
      threadOrder,
      hiddenThreads,
    }: Action));
  };
}

export function showThread(threadIndex: ThreadIndex): Action {
  return {
    type: 'SHOW_THREAD',
    threadIndex,
  };
}

export function changeCallTreeSearchString(searchString: string): Action {
  return {
    type: 'CHANGE_CALL_TREE_SEARCH_STRING',
    searchString,
  };
}

export function changeExpandedFuncStacks(
  threadIndex: ThreadIndex,
  expandedFuncStacks: Array<IndexIntoFuncTable[]>
): Action {
  return {
    type: 'CHANGE_EXPANDED_FUNC_STACKS',
    threadIndex, expandedFuncStacks,
  };
}

export function changeSelectedMarker(
  threadIndex: ThreadIndex,
  selectedMarker: IndexIntoMarkersTable | -1
): Action {
  return {
    type: 'CHANGE_SELECTED_MARKER',
    selectedMarker, threadIndex,
  };
}

export function changeImplementationFilter(implementation: ImplementationFilter): Action {
  return {
    type: 'CHANGE_IMPLEMENTATION_FILTER',
    implementation,
  };
}

export function changeInvertCallstack(invertCallstack: boolean): Action {
  return {
    type: 'CHANGE_INVERT_CALLSTACK',
    invertCallstack,
  };
}

export function changeHidePlatformDetails(hidePlatformDetails: boolean): Action {
  return {
    type: 'CHANGE_HIDE_PLATFORM_DETAILS',
    hidePlatformDetails,
  };
}

export type UpdateProfileSelection = (selection: ProfileSelection) => Action;

export function updateProfileSelection(selection: ProfileSelection): Action {
  return {
    type: 'UPDATE_PROFILE_SELECTION',
    selection,
  };
}

export function addRangeFilter(start: number, end: number): Action {
  return {
    type: 'ADD_RANGE_FILTER',
    start, end,
  };
}

export function addRangeFilterAndUnsetSelection(start: number, end: number): ThunkAction<void> {
  return dispatch => {
    dispatch(addRangeFilter(start, end));
    dispatch(updateProfileSelection({ hasSelection: false, isModifying: false }));
  };
}

export function popRangeFilters(firstRemovedFilterIndex: number): Action {
  return {
    type: 'POP_RANGE_FILTERS',
    firstRemovedFilterIndex,
  };
}

export function popRangeFiltersAndUnsetSelection(firstRemovedFilterIndex: number): ThunkAction<void> {
  return dispatch => {
    dispatch(popRangeFilters(firstRemovedFilterIndex));
    dispatch(updateProfileSelection({ hasSelection: false, isModifying: false }));
  };
}

export function addCallTreeFilter(threadIndex: ThreadIndex, filter: CallTreeFilter): Action {
  return {
    type: 'ADD_CALL_TREE_FILTER',
    threadIndex,
    filter,
  };
}

export function popCallTreeFilters(threadIndex: ThreadIndex, firstRemovedFilterIndex: number): Action {
  return {
    type: 'POP_CALL_TREE_FILTERS',
    threadIndex,
    firstRemovedFilterIndex,
  };
}
