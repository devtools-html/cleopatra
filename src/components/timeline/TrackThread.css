/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.timelineTrackThread {
  position: relative;
  z-index: 1;
  display: flex;
  flex: 1;
  flex-flow: column nowrap;
}

.timelineTrackThreadBlank {
  /* Clip off the border to the left, by overflowing by one pixel. */
  position: relative;
  left: -1px;
  width: 100%;
  height: 30px;
  padding-left: 1px;
  background-color: #fff;
}

.timelineTrackThreadMarkerOverview {
  position: relative;
  display: block;
  overflow: hidden;
  height: 6px;
  margin: 0;
  list-style: none;
  opacity: 0.75;
}

.timelineTrackThreadMarkerOverviewContainerJank {
  border-bottom: 1px solid var(--grey-30);
}

.timelineTrackThreadMarkerOverview.selected {
  opacity: 1;
}

.timelineTrackThreadMarkerOverviewThreadGeckoMain {
  height: 18px;
}
