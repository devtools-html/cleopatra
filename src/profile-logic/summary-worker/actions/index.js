/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import { summarizeProfile } from '../../../profile-logic/summarize-profile';

export function processProfileSummary() {
  return function (dispatch, getState) {
    dispatch({
      toContent: true,
      type: 'PROFILE_SUMMARY_PROCESSED',
      summary: summarizeProfile(getState().profile),
    });
  };
}

export function profileProcessed(profile) {
  return {
    type: 'PROFILE_PROCESSED',
    profile: profile,
  };
}
