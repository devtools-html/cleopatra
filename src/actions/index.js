/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as app from './app';
import * as icons from './icons';
import * as profileSummary from './profile-summary';
import * as profileView from './profile-view';
import * as receiveProfile from './receive-profile';
import * as timeline from './timeline';
import * as profileUpload from './profile-upload';
import * as shortUrl from './short-url';
import * as urlState from './url-state';

export default Object.assign(
  {},
  app,
  icons,
  profileSummary,
  profileUpload,
  profileView,
  receiveProfile,
  shortUrl,
  timeline,
  urlState
);
