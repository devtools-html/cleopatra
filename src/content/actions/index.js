// @flow
import * as app from './app';
import * as profileSummary from './profile-summary';
import * as profileView from './profile-view';
import * as receiveProfile from './receive-profile';
import * as timeline from './timeline';

export default Object.assign({}, app, profileSummary, profileView, receiveProfile, timeline);
