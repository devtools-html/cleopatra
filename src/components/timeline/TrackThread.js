/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React, { PureComponent } from 'react';
import classNames from 'classnames';
import memoize from 'memoize-immutable';
import explicitConnect from 'firefox-profiler/utils/connect';
import {
  withSize,
  type SizeProps,
} from 'firefox-profiler/components/shared/WithSize';
import { ThreadStackGraph } from 'firefox-profiler/components/shared/thread/StackGraph';
import { ThreadCPUGraph } from 'firefox-profiler/components/shared/thread/CPUGraph';
import { ThreadSampleGraph } from 'firefox-profiler/components/shared/thread/SampleGraph';
import { ThreadActivityGraph } from 'firefox-profiler/components/shared/thread/ActivityGraph';

import {
  getProfileInterval,
  getCommittedRange,
  getCategories,
  getSelectedThreadIndexes,
  getTimelineType,
  getInvertCallstack,
  getTimelineTrackOrganization,
  getThreadSelectorsFromThreadsKey,
  getMaxThreadCPUDelta,
  getIsExperimentalCPUGraphsEnabled,
} from 'firefox-profiler/selectors';
import {
  TimelineMarkersJank,
  TimelineMarkersFileIo,
  TimelineMarkersOverview,
  TimelineMarkersMemory,
} from './Markers';
import {
  updatePreviewSelection,
  changeRightClickedTrack,
  changeSelectedCallNode,
  focusCallTree,
  selectLeafCallNode,
  selectRootCallNode,
} from 'firefox-profiler/actions/profile-view';
import { reportTrackThreadHeight } from 'firefox-profiler/actions/app';
import { hasThreadKeys } from 'firefox-profiler/profile-logic/profile-data';
import { EmptyThreadIndicator } from './EmptyThreadIndicator';
import { getTrackSelectionModifier } from 'firefox-profiler/utils';
import { assertExhaustiveCheck } from 'firefox-profiler/utils/flow';
import './TrackThread.css';

import type {
  TimelineType,
  Thread,
  ThreadIndex,
  CategoryList,
  IndexIntoSamplesTable,
  Milliseconds,
  StartEndRange,
  CallNodeInfo,
  IndexIntoCallNodeTable,
  SelectedState,
  State,
  TimelineTrackOrganization,
  ThreadsKey,
} from 'firefox-profiler/types';

import type { ConnectedProps } from 'firefox-profiler/utils/connect';

type OwnProps = {|
  +threadsKey: ThreadsKey,
  +trackType: 'expanded' | 'condensed',
  +showMemoryMarkers?: boolean,
  +trackName: string,
|};

type StateProps = {|
  +fullThread: Thread,
  +rangeFilteredThread: Thread,
  +filteredThread: Thread,
  +tabFilteredThread: Thread,
  +callNodeInfo: CallNodeInfo,
  +selectedCallNodeIndex: IndexIntoCallNodeTable | null,
  +unfilteredSamplesRange: StartEndRange | null,
  +interval: Milliseconds,
  +rangeStart: Milliseconds,
  +rangeEnd: Milliseconds,
  +sampleIndexOffset: number,
  +categories: CategoryList,
  +timelineType: TimelineType,
  +hasFileIoMarkers: boolean,
  +samplesSelectedStates: null | SelectedState[],
  +invertCallstack: boolean,
  +treeOrderSampleComparator: (
    IndexIntoSamplesTable,
    IndexIntoSamplesTable
  ) => number,
  +timelineTrackOrganization: TimelineTrackOrganization,
  +selectedThreadIndexes: Set<ThreadIndex>,
  +enableCPUUsage: boolean,
  +isExperimentalCPUGraphsEnabled: boolean,
  +maxThreadCPUDelta: number,
|};

type DispatchProps = {|
  +changeRightClickedTrack: typeof changeRightClickedTrack,
  +updatePreviewSelection: typeof updatePreviewSelection,
  +changeSelectedCallNode: typeof changeSelectedCallNode,
  +focusCallTree: typeof focusCallTree,
  +selectLeafCallNode: typeof selectLeafCallNode,
  +selectRootCallNode: typeof selectRootCallNode,
  +reportTrackThreadHeight: typeof reportTrackThreadHeight,
|};

type Props = {|
  ...SizeProps,
  ...ConnectedProps<OwnProps, StateProps, DispatchProps>,
|};

class TimelineTrackThreadImpl extends PureComponent<Props> {
  /**
   * Handle when a sample is clicked in the ThreadStackGraph and in the ThreadActivityGraph.
   * This will select the leaf-most stack frame or call node.
   */
  _onSampleClick = (
    event: SyntheticMouseEvent<>,
    sampleIndex: IndexIntoSamplesTable
  ) => {
    const modifier = getTrackSelectionModifier(event);
    switch (modifier) {
      case 'none': {
        const {
          threadsKey,
          selectLeafCallNode,
          selectRootCallNode,
          focusCallTree,
          invertCallstack,
          selectedThreadIndexes,
        } = this.props;

        // Sample clicking only works for one thread. See issue #2709
        if (selectedThreadIndexes.size === 1) {
          if (invertCallstack) {
            // When we're displaying the inverted call stack, the "leaf" call node we're
            // interested in is actually displayed as the "root" of the tree.
            selectRootCallNode(threadsKey, sampleIndex);
          } else {
            selectLeafCallNode(threadsKey, sampleIndex);
          }
          focusCallTree();
        }
        if (
          typeof threadsKey === 'number' &&
          selectedThreadIndexes.has(threadsKey)
        ) {
          // We could have multiple threads selected here, and we wouldn't want
          // to de-select one when interacting with it.
          event.stopPropagation();
        }
        break;
      }
      case 'ctrl':
        // Do nothing, the track selection logic will kick in.
        break;
      default:
        assertExhaustiveCheck(modifier, 'Unhandled modifier case.');
        break;
    }
  };

  _onMarkerSelect = (start: Milliseconds, end: Milliseconds) => {
    const { rangeStart, rangeEnd, updatePreviewSelection } = this.props;
    updatePreviewSelection({
      hasSelection: true,
      isModifying: false,
      selectionStart: Math.max(rangeStart, start),
      selectionEnd: Math.min(rangeEnd, end),
    });
  };

  componentDidUpdate() {
    const { threadsKey, height, reportTrackThreadHeight } = this.props;
    // Most likely this track height shouldn't change, but if it does, report it.
    // The action will only dispatch on changed values.
    reportTrackThreadHeight(threadsKey, height);
  }

  render() {
    const {
      fullThread,
      filteredThread,
      rangeFilteredThread,
      tabFilteredThread,
      threadsKey,
      interval,
      rangeStart,
      rangeEnd,
      sampleIndexOffset,
      callNodeInfo,
      selectedCallNodeIndex,
      unfilteredSamplesRange,
      categories,
      timelineType,
      hasFileIoMarkers,
      showMemoryMarkers,
      samplesSelectedStates,
      treeOrderSampleComparator,
      trackType,
      timelineTrackOrganization,
      trackName,
      enableCPUUsage,
      maxThreadCPUDelta,
      isExperimentalCPUGraphsEnabled,
    } = this.props;

    const processType = filteredThread.processType;
    const displayJank = processType !== 'plugin';
    const displayMarkers =
      (filteredThread.name === 'GeckoMain' ||
        filteredThread.name === 'Compositor' ||
        filteredThread.name === 'Renderer' ||
        filteredThread.name === 'Java Main Thread' ||
        filteredThread.name === 'Merged thread' ||
        filteredThread.name.startsWith('MediaDecoderStateMachine')) &&
      processType !== 'plugin';

    return (
      <div className={classNames('timelineTrackThread', trackType)}>
        {timelineTrackOrganization.type !== 'active-tab' ? (
          <>
            {showMemoryMarkers ? (
              <TimelineMarkersMemory
                rangeStart={rangeStart}
                rangeEnd={rangeEnd}
                threadsKey={threadsKey}
                onSelect={this._onMarkerSelect}
              />
            ) : null}
            {hasFileIoMarkers ? (
              <TimelineMarkersFileIo
                rangeStart={rangeStart}
                rangeEnd={rangeEnd}
                threadsKey={threadsKey}
                onSelect={this._onMarkerSelect}
              />
            ) : null}
            {displayJank ? (
              <TimelineMarkersJank
                rangeStart={rangeStart}
                rangeEnd={rangeEnd}
                threadsKey={threadsKey}
                onSelect={this._onMarkerSelect}
              />
            ) : null}
            {displayMarkers ? (
              <TimelineMarkersOverview
                rangeStart={rangeStart}
                rangeEnd={rangeEnd}
                threadsKey={threadsKey}
                onSelect={this._onMarkerSelect}
              />
            ) : null}
          </>
        ) : null}
        {(timelineType === 'category' || timelineType === 'cpu-category') &&
        !filteredThread.isJsTracer ? (
          <>
            <ThreadActivityGraph
              className="threadActivityGraph"
              trackName={trackName}
              interval={interval}
              fullThread={fullThread}
              rangeFilteredThread={rangeFilteredThread}
              rangeStart={rangeStart}
              rangeEnd={rangeEnd}
              sampleIndexOffset={sampleIndexOffset}
              onSampleClick={this._onSampleClick}
              categories={categories}
              samplesSelectedStates={samplesSelectedStates}
              treeOrderSampleComparator={treeOrderSampleComparator}
              enableCPUUsage={enableCPUUsage}
              maxThreadCPUDelta={maxThreadCPUDelta}
            />
            <ThreadSampleGraph
              className="threadSampleGraph"
              trackName={trackName}
              interval={interval}
              thread={filteredThread}
              tabFilteredThread={tabFilteredThread}
              rangeStart={rangeStart}
              rangeEnd={rangeEnd}
              callNodeInfo={callNodeInfo}
              selectedCallNodeIndex={selectedCallNodeIndex}
              categories={categories}
              onSampleClick={this._onSampleClick}
            />
            {isExperimentalCPUGraphsEnabled &&
            rangeFilteredThread.samples.threadCPUDelta !== undefined ? (
              <ThreadCPUGraph
                className="threadCPUGraph"
                trackName={trackName}
                interval={interval}
                thread={filteredThread}
                tabFilteredThread={tabFilteredThread}
                rangeStart={rangeStart}
                rangeEnd={rangeEnd}
                callNodeInfo={callNodeInfo}
                selectedCallNodeIndex={selectedCallNodeIndex}
                categories={categories}
                onSampleClick={this._onSampleClick}
                maxThreadCPUDelta={maxThreadCPUDelta}
              />
            ) : null}
          </>
        ) : (
          <ThreadStackGraph
            className="threadStackGraph"
            trackName={trackName}
            interval={interval}
            thread={filteredThread}
            tabFilteredThread={tabFilteredThread}
            rangeStart={rangeStart}
            rangeEnd={rangeEnd}
            callNodeInfo={callNodeInfo}
            selectedCallNodeIndex={selectedCallNodeIndex}
            categories={categories}
            onSampleClick={this._onSampleClick}
          />
        )}
        {timelineTrackOrganization.type === 'active-tab' ? (
          <div className="timelineTrackThreadMarkers">
            {trackType === 'expanded' && showMemoryMarkers ? (
              <TimelineMarkersMemory
                rangeStart={rangeStart}
                rangeEnd={rangeEnd}
                threadsKey={threadsKey}
                onSelect={this._onMarkerSelect}
              />
            ) : null}
            {trackType === 'expanded' && hasFileIoMarkers ? (
              <TimelineMarkersFileIo
                rangeStart={rangeStart}
                rangeEnd={rangeEnd}
                threadsKey={threadsKey}
                onSelect={this._onMarkerSelect}
              />
            ) : null}
            {displayJank ? (
              <TimelineMarkersJank
                rangeStart={rangeStart}
                rangeEnd={rangeEnd}
                threadsKey={threadsKey}
                onSelect={this._onMarkerSelect}
              />
            ) : null}
            {trackType === 'expanded' && displayMarkers ? (
              <TimelineMarkersOverview
                rangeStart={rangeStart}
                rangeEnd={rangeEnd}
                threadsKey={threadsKey}
                onSelect={this._onMarkerSelect}
              />
            ) : null}
          </div>
        ) : null}
        <EmptyThreadIndicator
          thread={filteredThread}
          interval={interval}
          rangeStart={rangeStart}
          rangeEnd={rangeEnd}
          unfilteredSamplesRange={unfilteredSamplesRange}
        />
      </div>
    );
  }
}

/**
 * Memoize the hasThreadKeys to not compute it all the time.
 */
const _getTimelineIsSelected = memoize(
  (selectedThreads, threadsKey) => hasThreadKeys(selectedThreads, threadsKey),
  { limit: 1 }
);

export const TimelineTrackThread = explicitConnect<
  OwnProps,
  StateProps,
  DispatchProps
>({
  mapStateToProps: (state: State, ownProps: OwnProps) => {
    const { threadsKey } = ownProps;
    const selectors = getThreadSelectorsFromThreadsKey(threadsKey);
    const selectedThreadIndexes = getSelectedThreadIndexes(state);
    const committedRange = getCommittedRange(state);
    const selectedCallNodeIndex = _getTimelineIsSelected(
      selectedThreadIndexes,
      threadsKey
    )
      ? selectors.getSelectedCallNodeIndex(state)
      : null;
    const fullThread = selectors.getCPUProcessedThread(state);
    const timelineType = getTimelineType(state);
    const enableCPUUsage =
      timelineType === 'cpu-category' &&
      fullThread.samples.threadCPUDelta !== undefined;

    return {
      invertCallstack: getInvertCallstack(state),
      fullThread,
      filteredThread: selectors.getFilteredThread(state),
      rangeFilteredThread: selectors.getRangeFilteredThread(state),
      tabFilteredThread: selectors.getTabFilteredThread(state),
      callNodeInfo: selectors.getCallNodeInfo(state),
      selectedCallNodeIndex,
      unfilteredSamplesRange: selectors.unfilteredSamplesRange(state),
      interval: getProfileInterval(state),
      rangeStart: committedRange.start,
      rangeEnd: committedRange.end,
      sampleIndexOffset: selectors.getSampleIndexOffsetFromCommittedRange(
        state
      ),
      categories: getCategories(state),
      timelineType,
      hasFileIoMarkers:
        selectors.getTimelineFileIoMarkerIndexes(state).length !== 0,
      samplesSelectedStates: selectors.getSamplesSelectedStatesInFilteredThread(
        state
      ),
      treeOrderSampleComparator: selectors.getTreeOrderComparatorInFilteredThread(
        state
      ),
      timelineTrackOrganization: getTimelineTrackOrganization(state),
      selectedThreadIndexes,
      enableCPUUsage,
      isExperimentalCPUGraphsEnabled: getIsExperimentalCPUGraphsEnabled(state),
      maxThreadCPUDelta: getMaxThreadCPUDelta(state),
    };
  },
  mapDispatchToProps: {
    updatePreviewSelection,
    changeRightClickedTrack,
    changeSelectedCallNode,
    focusCallTree,
    selectLeafCallNode,
    selectRootCallNode,
    reportTrackThreadHeight,
  },
  component: withSize<Props>(TimelineTrackThreadImpl),
});
