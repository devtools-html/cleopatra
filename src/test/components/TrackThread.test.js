/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import type { Profile } from '../../types/profile';

import * as React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';

import { changeTimelineType } from '../../actions/profile-view';
import TrackThread from '../../components/timeline/TrackThread';
import { getPreviewSelection } from '../../selectors/profile';
import { selectedThreadSelectors } from '../../selectors/per-thread';
import mockCanvasContext from '../fixtures/mocks/canvas-context';
import mockRaf from '../fixtures/mocks/request-animation-frame';
import { storeWithProfile } from '../fixtures/stores';
import {
  getBoundingBox,
  getMouseEvent,
  addRootOverlayElement,
  removeRootOverlayElement,
} from '../fixtures/utils';

import {
  getProfileFromTextSamples,
  getProfileWithMarkers,
} from '../fixtures/profiles/make-profile';

// The graph is 400 pixels wide based on the getBoundingBox mock. Each stack is 100
// pixels wide. Use the value 50 to click in the middle of this stack, and
// incrementing by steps of 100 pixels to get to the next stack.
const GRAPH_WIDTH = 400;
const GRAPH_HEIGHT = 50;
const STACK_1_X_POSITION = 50;
const STACK_2_X_POSITION = 150;
const STACK_3_X_POSITION = 250;
const STACK_4_X_POSITION = 350;

/**
 * This test is asserting behavior more for the ThreadStackGraph component. The
 * ThreadActivityGraph component was added as a new default. Currently this test
 * only checks the older behavior.
 */
describe('timeline/TrackThread', function() {
  beforeEach(addRootOverlayElement);
  afterEach(removeRootOverlayElement);

  function getSamplesProfile() {
    return getProfileFromTextSamples(`
      a  d  g  j
      b  e  h  k
      c  f  i  l
    `).profile;
  }

  function getMarkersProfile() {
    const profile = getProfileWithMarkers([
      ['Marker A', 0, { startTime: 0, endTime: 1 }],
      ['Marker B', 1, { startTime: 1, endTime: 2 }],
      ['Marker C', 2, { startTime: 2, endTime: 3 }],
      ['Marker D', 3, { startTime: 3, endTime: 4 }],
    ]);
    const [thread] = profile.threads;
    thread.name = 'GeckoMain';
    thread.processType = 'default';
    return profile;
  }

  function setup(profile: Profile) {
    const store = storeWithProfile(profile);
    const { getState, dispatch } = store;
    const threadIndex = 0;
    const flushRafCalls = mockRaf();
    const ctx = mockCanvasContext();
    jest
      .spyOn(HTMLCanvasElement.prototype, 'getContext')
      .mockImplementation(() => ctx);

    jest
      .spyOn(HTMLElement.prototype, 'getBoundingClientRect')
      .mockImplementation(() => getBoundingBox(GRAPH_WIDTH, GRAPH_HEIGHT));

    // Note: These tests were first written with the timeline using the ThreadStackGraph.
    // This is not the default view, so dispatch an action to change to the older default
    // view.
    store.dispatch(changeTimelineType('stack'));

    const view = mount(
      <Provider store={store}>
        <TrackThread threadIndex={threadIndex} />
      </Provider>
    );

    // WithSize uses requestAnimationFrame
    flushRafCalls();
    view.update();

    const stackGraphCanvas = view.find('.threadStackGraphCanvas').first();
    const tracingMarkersCanvas = view
      .find(
        [
          '.timelineTrackThreadIntervalMarkerOverviewThreadGeckoMain',
          '.timelineTracingMarkersCanvas',
        ].join(' ')
      )
      .first();

    return {
      dispatch,
      getState,
      profile,
      thread: profile.threads[0],
      store,
      view,
      threadIndex,
      stackGraphCanvas,
      tracingMarkersCanvas,
    };
  }

  it('matches the snapshot', () => {
    const { view } = setup(getSamplesProfile());
    expect(view).toMatchSnapshot();
  });

  it('has the correct selectors into useful parts of the component for the samples profile', function() {
    const { stackGraphCanvas } = setup(getSamplesProfile());
    expect(stackGraphCanvas.exists()).toBe(true);
  });

  it('has the correct selectors into useful parts of the component for the markers profile', function() {
    const { tracingMarkersCanvas } = setup(getMarkersProfile());
    expect(tracingMarkersCanvas.exists()).toBe(true);
  });

  it('can click a stack in the stack graph', function() {
    const { getState, stackGraphCanvas, thread } = setup(getSamplesProfile());

    // Provide a quick helper for nicely asserting the call node path.
    const getCallNodePath = () =>
      selectedThreadSelectors
        .getSelectedCallNodePath(getState())
        .map(funcIndex =>
          thread.stringTable.getString(thread.funcTable.name[funcIndex])
        );

    stackGraphCanvas.simulate(
      'mouseup',
      getMouseEvent({ pageX: STACK_1_X_POSITION })
    );
    expect(getCallNodePath()).toEqual(['a', 'b', 'c']);

    stackGraphCanvas.simulate(
      'mouseup',
      getMouseEvent({ pageX: STACK_2_X_POSITION })
    );
    expect(getCallNodePath()).toEqual(['d', 'e', 'f']);

    stackGraphCanvas.simulate(
      'mouseup',
      getMouseEvent({ pageX: STACK_3_X_POSITION })
    );
    expect(getCallNodePath()).toEqual(['g', 'h', 'i']);

    stackGraphCanvas.simulate(
      'mouseup',
      getMouseEvent({ pageX: STACK_4_X_POSITION })
    );
    expect(getCallNodePath()).toEqual(['j', 'k', 'l']);
  });

  it('can click a marker', function() {
    const { getState, tracingMarkersCanvas } = setup(getMarkersProfile());

    function clickAndGetMarkerName(pageX: number) {
      tracingMarkersCanvas.simulate('mousedown', getMouseEvent({ pageX }));
      tracingMarkersCanvas.simulate('mouseup', getMouseEvent({ pageX }));
      return getPreviewSelection(getState());
    }

    expect(clickAndGetMarkerName(STACK_1_X_POSITION)).toMatchObject({
      selectionStart: 0,
      selectionEnd: 1,
    });

    expect(clickAndGetMarkerName(STACK_2_X_POSITION)).toMatchObject({
      selectionStart: 1,
      selectionEnd: 2,
    });
  });
});
