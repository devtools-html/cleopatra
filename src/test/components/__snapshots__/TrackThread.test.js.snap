// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`timeline/TrackThread matches the snapshot 1`] = `
<div
  class="timelineTrackThread"
>
  <div
    class="timelineMarkers selected"
    data-testid="TimelineMarkersJank"
  >
    <canvas
      class="timelineMarkersCanvas"
    />
  </div>
  <div
    class="threadStackGraph"
  >
    <canvas
      class="threadStackGraphCanvas threadStackGraphCanvas"
      height="50"
      width="400"
    />
  </div>
  <div
    class="timelineEmptyThreadIndicator"
  />
</div>
`;
